### PR TITLE
dev-embedded/jal: Fix incompatible integer to pointer conversion

### PR DESCRIPTION
**Pull request by listout**
Closes: https://bugs.gentoo.org/883243

### CODE REVIEW

**Comment by listout** [comment]
Upstream seems dead, last updated 8 years ago.

File: dev-embedded/jal/files/jal-0.4.62-fix-incompatible-integer-to-pointer-clang16.patch [hunk 1]
```diff
@@ -0,0 +1,24 @@
+Bug: https://bugs.gentoo.org/883243
+--- a/jal/treerep.c
++++ b/jal/treerep.c
+@@ -892,7 +892,7 @@ tree new_asm(loc_t loc, opcode_t opcode, tree arg1, int arg2)
+ 
+ 
+     if (code_has(p->opcode, field_flabel)) {
+-        assert_pointer(NULL, arg2);
++        assert_pointer(NULL, &arg2);
```
Comment (ceamac) [inline dev-embedded/jal/files/jal-0.4.62-fix-incompatible-integer-to-pointer-clang16.patch:9]: This is wrong.  By looking at the code, `assert_pointer` seems to do some pointer arithmetic to determine if the given value is inside some custom allocation block; `&arg2` points to the stack, so the assert will always fail.  I suppose it expects arg2 to be a pointer as int, so a simple cast should be sufficient in this case.

File: dev-embedded/jal/files/jal-0.4.62-fix-incompatible-integer-to-pointer-clang16.patch [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ 
+--- a/jal/treetools.c
++++ b/jal/treetools.c
+@@ -600,7 +600,8 @@ tree arg(tree p, int n, boolean must_find)
+ #endif
+     assert_kind(p->loc, p, node_procedure);
+ /*    jal_assert(p->loc, n > 0); */
+-    assert_pointer(p->loc, n > 0);
++	int temp_val = (n > 1);
++    assert_pointer(p->loc, &temp_val);
```
Comment (ceamac) [inline dev-embedded/jal/files/jal-0.4.62-fix-incompatible-integer-to-pointer-clang16.patch:21]: This is also wrong, but by the looks of it the original is also wrong.  I don't know why they commented out `jal_assert`, it looks like it should have been correct in the first place.  Probably it didn't work?

**Review by ceamac — COMMENTED** [review]
Unfortunately there are no tests, the codebase is quite old, upstream is unresponsive, probably gone, and according to repology Gentoo is the only distribution carrying this package.

I don't know how to check the fix in `jal/treetools.c`, even if we did one.

**Comment by listout** [comment]
@ceamac Thank you very much for the review. I've modified my patch, can you review once again.

**Pull request closed without merging**